pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,316 @@
+	dataHandler        DataHandlerFunc
+	dataHandlerCloseCh chan chan struct{}
+	dataWrittenCh      chan bool
+	optionCallback     func(byte, byte)
+	connReadDoneCh     chan chan struct{}
+	connWriteDoneCh    chan chan struct{}
+	negotiationDone    chan struct{}
+	closedMutex        sync.Mutex
+	closed             bool
+}
+
+// Safely read/write concurrently to the data Buffer
+// databuffer is written to by the FSM and it is read from by the dataHandler
+type dataReadWriter struct {
+	dataBuffer bytes.Buffer
+	dataMux    *sync.Mutex
+}
+
+func (drw *dataReadWriter) Read(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Read(p)
+}
+
+func (drw *dataReadWriter) Write(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Write(p)
+}
+
+type connectionWriter struct {
+	ch chan []byte
+}
+
+func (cw *connectionWriter) Write(b []byte) (int, error) {
+	cw.ch <- b
+	return len(b), nil
+}
+
+func newTelnetConn(opts connOpts) *TelnetConn {
+	tc := &TelnetConn{
+		conn:               opts.conn,
+		readCh:             make(chan []byte),
+		writeCh:            make(chan []byte),
+		unackedServerOpts:  make(map[byte]bool),
+		unackedClientOpts:  make(map[byte]bool),
+		cmdHandler:         opts.cmdHandler,
+		dataHandler:        opts.dataHandler,
+		dataHandlerCloseCh: make(chan chan struct{}),
+		dataWrittenCh:      make(chan bool),
+		serverOpts:         opts.serverOpts,
+		clientOpts:         opts.clientOpts,
+		optionCallback:     opts.optCallback,
+		fsmInputCh:         make(chan byte),
+		connReadDoneCh:     make(chan chan struct{}),
+		connWriteDoneCh:    make(chan chan struct{}),
+		negotiationDone:    make(chan struct{}),
+		closed:             false,
+	}
+	if tc.optionCallback == nil {
+		tc.optionCallback = tc.handleOptionCommand
+	}
+	tc.handlerWriter = &connectionWriter{
+		ch: tc.writeCh,
+	}
+	tc.dataRW = &dataReadWriter{
+		dataMux: new(sync.Mutex),
+	}
+	fsm := opts.fsm
+	fsm.tc = tc
+	tc.fsm = fsm
+	for k := range tc.serverOpts {
+		tc.unackedServerOpts[k] = true
+	}
+	for k := range tc.clientOpts {
+		tc.unackedClientOpts[k] = true
+	}
+	return tc
+}
+
+func (c *TelnetConn) connectionLoop() {
+	log.Debugf("Entered connectionLoop")
+	// this is the reading thread
+	go func() {
+		for {
+			select {
+			case readBytes := <-c.readCh:
+				for _, ch := range readBytes {
+					c.fsmInputCh <- ch
+				}
+
+			case ch := <-c.connReadDoneCh:
+				ch <- struct{}{}
+				return
+			}
+		}
+	}()
+	// this is the writing thread
+	go func() {
+		for {
+			select {
+			case writeBytes := <-c.writeCh:
+				c.conn.Write(writeBytes)
+			case ch := <-c.connWriteDoneCh:
+				ch <- struct{}{}
+				return
+			}
+		}
+	}()
+}
+
+// reads from the connection and dumps into the connection read channel
+func (c *TelnetConn) readLoop() {
+	defer func() {
+		log.Debugf("read loop closed")
+	}()
+	for {
+		buf := make([]byte, 4096)
+		n, err := c.conn.Read(buf)
```

Comment:
You can turn `conn` into a buffered `io.Reader` using `bufio.NewReader`.  Up to you to use it, but it's there. :-)